### PR TITLE
Added support for AWS Sigv4 for UrlLib3.

### PR DESCRIPTION
**Pull request by dblock**
### Description

- Adds SigV4 support for Urllib3HttpConnection.
- Deprecates `AWSV4SignerAuth` in favor of `RequestsAWSV4SignerAuth`
- Uses `Urllib3AWSV4SignerAuth` in the documentation since that transport is faster.

#### AOS

```
export SERVICE=es
export ENDPOINT=https://xyz.us-west-2.aoss.amazonaws.com
poetry run python sigv4/urllib3-hello.py 

INFO:Found credentials in environment variables.
INFO:GET https://search-dblock-test-opensearch-27-xyz.us-west-2.es.amazonaws.com:443/ [status:200 request:0.768s]
opensearch: 2.7.0
INFO:PUT https://search-dblock-test-opensearch-27-xyz.us-west-2.es.amazonaws.com:443/movies [status:200 request:2.298s]
INFO:PUT https://search-dblock-test-opensearch-27-xyz.us-west-2.es.amazonaws.com:443/movies/_doc/1 [status:201 request:0.500s]
INFO:POST https://search-dblock-test-opensearch-27-xyz.us-west-2.es.amazonaws.com:443/_search [status:200 request:0.213s]
{'director': 'Bennett Miller', 'title': 'Moneyball', 'year': 2011}
INFO:DELETE https://search-dblock-test-opensearch-27-xyz.us-west-2.es.amazonaws.com:443/movies/_doc/1 [status:200 request:0.108s]
INFO:DELETE https://search-dblock-test-opensearch-27-xyz.us-west-2.es.amazonaws.com:443/movies [status:200 request:0.223s]
```

#### AOSS

```
export SERVICE=aoss
export ENDPOINT=https://xyz.us-west-2.aoss.amazonaws.com
poetry run python sigv4/urllib3-hello.py 
```

```
INFO:Found credentials in environment variables.
INFO:PUT https://xyz.us-west-2.aoss.amazonaws.com:443/movies [status:200 request:4.173s]
INFO:PUT https://xyz.us-west-2.aoss.amazonaws.com:443/movies/_doc/1 [status:201 request:3.117s]
INFO:POST https://xyz.us-west-2.aoss.amazonaws.com:443/_search [status:200 request:4.916s]
{'director': 'Bennett Miller', 'title': 'Moneyball', 'year': 2011}
INFO:DELETE https://xyz.us-west-2.aoss.amazonaws.com:443/movies/_doc/1 [status:200 request:1.249s]
INFO:DELETE https://xyz.us-west-2.aoss.amazonaws.com:443/movies [status:200 request:0.168s]
```

### Issues Resolved

Closes https://github.com/opensearch-project/opensearch-py/issues/546

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
@harshavamsi would appreciate your CR since you did support for the requests connection class.

File: test_opensearchpy/test_connection/test_urllib3_http_connection.py [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+#
+# Modifications Copyright OpenSearch Contributors. See
+# GitHub history for details.
+#
+#  Licensed to Elasticsearch B.V. under one or more contributor
```
Comment (VachaShah) [inline test_opensearchpy/test_connection/test_urllib3_http_connection.py:11]: Do we need this license here?

File: guides/auth.md [hunk 2]
```diff
@@ -9,24 +9,24 @@ OpenSearch allows you to use different methods for the authentication via `conne
 
 ## IAM Authentication
 
-Opensearch-py supports IAM-based authentication via `AWSV4SignerAuth`, which uses `RequestHttpConnection` as the transport class for communicating with OpenSearch clusters running in Amazon Managed OpenSearch and OpenSearch Serverless, and works in conjunction with [botocore](https://pypi.org/project/botocore/).
+Opensearch-py supports IAM-based authentication via `RequestsAWSV4SignerAuth` and `Urllib3AWSV4SignerAuth`, which use `RequestHttpConnection` and `Urllib3HttpConnection` respectively, as the transport classes for communicating with OpenSearch clusters running in Amazon Managed OpenSearch and OpenSearch Serverless, and works in conjunction with [botocore](https://pypi.org/project/botocore/).
```
Comment (VachaShah) [inline guides/auth.md:12]: Do we recommend `Urllib3AWSV4SignerAuth` here?

File: opensearchpy/helpers/signer.py [hunk 3]
```diff
@@ -101,9 +95,49 @@ def _sign_request(self, prepared_request):  # type: ignore
...
+class AWSV4SignerAuth(RequestsAWSV4SignerAuth):
+    pass
+
+
+class Urllib3AWSV4SignerAuth(Callable):  # type: ignore
+    def __init__(self, credentials, region, service="es"):  # type: ignore
+        self.signer = AWSV4Signer(credentials, region, service)
+
+    def __call__(self, method: str, url: str, body: Any) -> Dict[str, str]:
+        return self.signer.sign(method, url, body)
```
Comment (VachaShah) [inline opensearchpy/helpers/signer.py:143]: `RequestsAWSV4SignerAuth` returns prepared request while `Urllib3AWSV4SignerAuth` returns the headers, should we keep them consistent?

File: test_opensearchpy/test_connection/test_urllib3_http_connection.py (hunk 1 above)
Comment (dblock) [reply test_opensearchpy/test_connection/test_urllib3_http_connection.py:11]: Yes, this code was refactored out of `test_connection.py`, so let's keep it.

File: guides/auth.md (hunk 2 above)
Comment (dblock) [reply guides/auth.md]: I've clarified the documentation in https://github.com/opensearch-project/opensearch-py/pull/547/commits/cb768749be020ca3882815d73464bb5bf5dee6a7.

**Comment by dblock** [comment]
@VachaShah LMK if there's anything else needed for this one

File: opensearchpy/helpers/signer.py (hunk 3 above)
Comment (VachaShah) [reply opensearchpy/helpers/signer.py:142]: That makes sense, thank you!

**Review by VachaShah — APPROVED** [review]
LGTM!

**Pull request merged**